apps/src/calc/inputIterator.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * Given a set of values (i.e. [1,2,3], and a number of parameters, generates
+ * all possible combinations of values.
+ */
+var InputIterator = function (values, numParams) {
```

Comment:
Ah! OK, I initially misread this class, I see what it's doing now—generating all possible permutations for a given length, AND allowing repeated inputs.

One standard way of doing this appears to be taking a cartesian product of an array and itself a given number of times: http://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript

E.g., using the example from there (input 1,2,3, and you want 2 per, you use 2 [1,2,3]s):

```
JSON.stringify(cartProd([1,2,3], [1,2,3]))
"[[1,1],[1,2],[1,3],[2,1],[2,2],[2,3],[3,1],[3,2],[3,3]]"
```

And if you wanted 3 per product:

```
JSON.stringify(cartProd([1,2,3], [1,2,3], [1,2,3]))
"[[1,1,1],[1,1,2],[1,1,3],[1,2,1],[1,2,2],[1,2,3],[1,3,1],[1,3,2],[1,3,3],[2,1,1],[2,1,2],[2,1,3],[2,2,1],[2,2,2],[2,2,3],[2,3,1],[2,3,2],[2,3,3],[3,1,1],[3,1,2],[3,1,3],[3,2,1],[3,2,2],[3,2,3],[3,3,1],[3,3,2],[3,3,3]]"
```

But that does it all-at-once rather than in an iterator style, could be an option if that fits the use case.